src/librustc_mir/borrow_check/mod.rs

```diff
@@ -367,8 +379,10 @@ pub struct MirBorrowckCtxt<'cx, 'gcx: 'tcx, 'tcx: 'cx> {
     /// at the time we detect and report a reservation error.
     reservation_error_reported: FxHashSet<Place<'tcx>>,
     /// This field keeps track of errors reported in the checking of moved variables,
-    /// so that we don't report report seemingly duplicate errors.
+    /// so that we don't report seemingly duplicate errors.
     moved_error_reported: FxHashSet<Place<'tcx>>,
+    /// Errors to be reported buffer
+    errors_buffer: Vec<DiagnosticBuilder<'cx>>,
```

Comment:
So, I hate to mention this after y'all have been wrestling with lifetimes for so long, but it occurs to me that this PR would be much simpler if `DiagnosticBuilder` did not carry a `'cx` lifetime. Indeed, if we look at the definition of `DiagnosticBuilder`, we see that it is the combination of two things:

https://github.com/rust-lang/rust/blob/bd455ef165f110143b97106b4e7cd8c83cd21f03/src/librustc_errors/diagnostic_builder.rs#L23-L29

I believe that the `diagnostic` field contains everything related to display the error and so forth, and the `Handler` just has the links to the session, stdout, etc. I believe that @michaelwoerister introduced this precisely so that we can serialize the set of errors produced by a query then load them later and "replay" them (also cc @estebank, who can problem confirm this).

I was thinking that maybe we should add a method like

```rust
impl DiagnosticBuilder<'cx> {
    fn buffer(self, buffered: &mut Vec<Diagnostic>) -> Diagnostic {
        buffered.push(self.diagnostic);
    }
}
```

and then we can replace all those calls like `.emit()` with `.buffer()`. Moreover, the buffer vector doesn't carry any lifetime.

In order to emit later, you can use [`DiagnosticBuilder::new_diagnostic(tcx.sess.diagnostic(), d)`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html#method.new_diagnostic) .

Reply:
I'll look into this now